packages/dds/tree/src/feature-libraries/flex-map-tree/mapTreeNode.ts

```diff
@@ -0,0 +1,744 @@
+	 * @param location - the parentage of this node, if it is being created underneath an existing node and field, or undefined if not
+	 * @remarks This class (and its subclasses) should not be directly constructed outside of this module.
+	 * Instead, use {@link getOrCreateMapTreeNode} to create a MapTreeNode from a {@link MapTree}.
+	 */
+	public constructor(
+		public readonly schema: TSchema,
+		public readonly mapTree: MapTree,
+		private location: LocationInField | undefined,
+	) {
+		assert(!nodeCache.has(mapTree), "A node already exists for the given MapTree");
+		nodeCache.set(mapTree, this);
+
+		// Fully demand the tree to ensure that parent pointers are present and accurate on all nodes.
+		// When a MapTreeNode is constructed, its MapTree may contain nodes (anywhere below) that map (via the `nodeCache`) to pre-existing MapTreeNodes.
+		// Put another way, for a given MapTree, some ancestor MapTreeNode can be created after any number of its descendant MapTreeNodes already exist.
+		// In such a case, the spine of nodes between the descendant and ancestor need to exist in order for the ancestor to be able to walk upwards via the `parentField` property.
+		// This needs to happen for all MapTreeNodes that are descendants of the ancestor MapTreeNode.
+		// Demanding the entire tree is overkill to solve this problem since not all descendant MapTree nodes will have corresponding MapTreeNodes.
+		// However, demanding the full tree also lets us eagerly validate that there are no duplicate MapTrees (i.e. same MapTree object) anywhere in the tree.
+		this.walkTree();
+	}
+
+	public get type(): TreeNodeSchemaIdentifier {
+		return this.schema.name;
+	}
+
+	/**
+	 * Set this node's parentage (see {@link FlexTreeNode.parentField}).
+	 * @remarks A node may only be adopted to a new parent one time, though this method may be called idempotently with the same parent and index.
+	 */
+	public adopt(parent: MapTreeField<FlexAllowedTypes>, index: number): void {
+		assert(
+			this.location === undefined ||
+				(parent.isSameAs(this.location.parent) && index === this.location.index),
+			"Node may not be reparented",
+		);
+		this.location = { parent, index };
+	}
+
+	/**
+	 * The field this tree is in, and the index within that field.
+	 * @remarks If this node is unparented, this method will return the special {@link rootMapTreeField} as the parent.
+	 */
+	public get parentField(): LocationInField {
+		if (this.location === undefined) {
+			return {
+				parent: rootMapTreeField,
+				index: -1,
+			};
+		}
+
+		return this.location;
+	}
+
+	public is<TSchemaInner extends FlexTreeNodeSchema>(
+		schema: TSchemaInner,
+	): this is FlexTreeTypedNode<TSchemaInner> {
+		return (schema as unknown) === this.schema;
+	}
+
+	public tryGetField(key: FieldKey): MapTreeField<FlexAllowedTypes> | undefined {
+		const field = this.mapTree.fields.get(key);
+		if (field !== undefined && field.length > 0) {
```

Comment:
Might be worth a quick comment explaining the `field.length > 0` check here. I don't think it's immediately obvious.

Reply:
Done